### PR TITLE
Threading fixes for config layers

### PR DESCRIPTION
**Pull request by CookiePLMonster**
This PR refactors Config Layouts functionality a bit, adding a layer of thread safety to them, as well as improving overall clarity. This fixes several "random" crashes on switching the game on/off and accessing configuration menus at the same time.

1) First commit revisits layer manager in `Config` namespace to make it thread safe. API has been made a bit stricter (`Config::GetLayers` was thankfully never used, though!) and some implementation details have been moved from the header file to the .cpp file. Now configuration layers are stored in a `shared_ptr` and not an `unique_ptr`, which cleanly solves an issue of one thread reading those, with another thread deleting them on terminating the game.
On top of that, any functions accessing `s_layers` have been protected with a read-write lock. It might seem like I added some redundant scopes, but that's not the case - `InvokeConfigChangedCallbacks` **must not** be protected with this mutex, else deadlocks occur!
2) Second commit revisits `LayerMap`. ~Previously, it was defined as `std::map<ConfigLocation, std::optional<std::string>>` and presence of a `nullopt` value was treated equally to lack of a value.~ I was able to remove this redundancy by replacing some uses of `operator[]` with `find()`. This also allowed me to mask `Get` functions as const, which was not possible before due to an implicit construction in `operator[]`. Additionally, I believe this may be a memory/performance improvement, since config layers will now **not** store redundant, empty entries for options they were probed on.
EDIT: This has been partially reverted to again store optionals in a map, since it's needed to achieve correct behaviour of deleting config keys. Fixes around `operator[]` persisted.

### Races fixed

I was unable to reproduce this race in Release mode, most likely due to very specific timing required to reproduce it. However, I was able to reproduce it in Debug:
1) Open Graphics options.
2) Launch the game, and quickly terminate it.
3) Repeat 2) until Dolphin crashes.

There were numerous reasons for those crashes - either one thread was reading from a `Layer` which was already deleted by another thread (that's what usage of `std::shared_ptr` fixes), or map's iterators get corrupted because one thread was iterating through it while the other was adding/removing entries to it (that's what a read-write lock fixes).

**NOTE:** There seems to be at least one more race on starting the game (`GetDSPProcessor()` returning null and being used without any validation), but this will **not** be fixed by this PR. If I find a fix better than "add a null check beforehand", I'll submit another PR.

### CODE REVIEW

File: Source/Core/Common/Config/Config.cpp [hunk 1]
```diff
@@ -5,51 +5,61 @@
...
 #include <map>
+#include <shared_mutex>
 
 #include "Common/Config/Config.h"
 
 namespace Config
 {
+using Layers = std::map<LayerType, std::shared_ptr<Layer>>;
+
+static std::shared_mutex s_layersRWLock;
```
Comment (lioncash) [inline Source/Core/Common/Config/Config.cpp]:
```suggestion
static std::shared_mutex s_layers_rw_lock;
```

We use `snake_case` as our naming scheme in new code

File: Source/Core/Common/Config/Config.cpp [hunk 2]
```diff
@@ -5,51 +5,61 @@
...
-Layers* GetLayers()
-{
-  return &s_layers;
-}
-
-void AddLayer(std::unique_ptr<Layer> layer)
+void AddLayerInternal(std::shared_ptr<Layer> layer)
 {
-  s_layers[layer->GetLayer()] = std::move(layer);
+  const Config::LayerType layerType = layer->GetLayer();
```
Comment (lioncash) [inline Source/Core/Common/Config/Config.cpp]:
```suggestion
  const Config::LayerType layer_type = layer->GetLayer();
```

File: Source/Core/Common/Config/Config.cpp [hunk 3]
```diff
@@ -5,51 +5,61 @@
...
+  AddLayerInternal(std::make_shared<Layer>(std::move(loader)));
 }
 
-Layer* GetLayer(LayerType layer)
+std::shared_ptr<Layer> GetLayer(LayerType layer)
 {
-  if (!LayerExists(layer))
-    return nullptr;
-  return s_layers[layer].get();
+  std::shared_ptr<Layer> result;
```
Comment (lioncash) [inline Source/Core/Common/Config/Config.cpp]: This can be moved closer to its assignment.

File: Source/Core/Common/Config/Config.cpp [hunk 4]
```diff
@@ -129,13 +151,16 @@ const std::string& GetLayerName(LayerType layer)
...
+  std::shared_lock lock(s_layersRWLock);
+
   for (auto layer : SEARCH_ORDER)
   {
-    if (!LayerExists(layer))
-      continue;
-
-    if (GetLayer(layer)->Exists(config))
-      return layer;
+    auto it = s_layers.find(layer);
```
Comment (lioncash) [inline Source/Core/Common/Config/Config.cpp]:
```suggestion
    const auto it = s_layers.find(layer);
```

File: Source/Core/Common/Config/Layer.cpp [hunk 5]
```diff
@@ -40,24 +40,20 @@ Layer::~Layer()
...
+  return iter != m_map.end();
 }
 
 bool Layer::DeleteKey(const ConfigLocation& location)
 {
   m_is_dirty = true;
-  bool had_value = m_map[location].has_value();
-  m_map[location].reset();
-  return had_value;
+  size_t elements_removed = m_map.erase(location);
```
Comment (lioncash) [inline Source/Core/Common/Config/Layer.cpp]:
```suggestion
  const size_t elements_removed = m_map.erase(location);
```

File: Source/Core/Common/Config/Layer.h [hunk 6]
```diff
@@ -103,18 +103,18 @@ class Layer
...
     return Get<T>(config_info.location).value_or(config_info.default_value);
   }
 
   template <typename T>
-  std::optional<T> Get(const ConfigLocation& location)
+  std::optional<T> Get(const ConfigLocation& location) const
   {
-    const std::optional<std::string>& str_value = m_map[location];
-    if (!str_value)
+    auto it = m_map.find(location);
```
Comment (lioncash) [inline Source/Core/Common/Config/Layer.h]:
```suggestion
    const auto it = m_map.find(location);
```

File: Source/Core/Common/Config/Layer.h [hunk 7]
```diff
@@ -129,13 +129,13 @@ class Layer
     Set(location, ValueToString(value));
   }
 
-  void Set(const ConfigLocation& location, const std::string& new_value)
+  void Set(const ConfigLocation& location, std::string new_value)
   {
-    std::optional<std::string>& current_value = m_map[location];
-    if (current_value == new_value)
+    auto it = m_map.find(location);
```
Comment (lioncash) [inline Source/Core/Common/Config/Layer.h]:
```suggestion
    const auto it = m_map.find(location);
```

File: Source/Core/Common/Config/Layer.h [hunk 8]
```diff
@@ -129,13 +129,13 @@ class Layer
...
+    if (it != m_map.end() && it->second == new_value)
       return;
     m_is_dirty = true;
-    current_value = new_value;
+    m_map[location] = std::move(new_value);
```
Comment (lioncash) [inline Source/Core/Common/Config/Layer.h]:
```suggestion
    m_map.insert_or_assign(location, std::move(new_value));
```

performs the same behavior without default constructing a temporary std::string if it doesn't exist in the map.

File: Source/Core/Common/Config/Config.cpp [hunk 9]
```diff
@@ -5,51 +5,78 @@
...
+std::shared_ptr<Layer> GetLayer(LayerType layer)
 {
-  if (!LayerExists(layer))
-    return nullptr;
-  return s_layers[layer].get();
+  std::shared_ptr<Layer> result;
+
+  ReadLock lock(s_layersRWLock);
+
+  auto it = s_layers.find(layer);
```
Comment (lioncash) [inline Source/Core/Common/Config/Config.cpp]:
```suggestion
  const auto it = s_layers.find(layer);
```

File: Source/Core/Common/Config/Config.cpp [hunk 10]
```diff
@@ -5,51 +5,78 @@
...
+using WriteLock = std::unique_lock<std::shared_mutex>;
+#endif
+
+void AddLayerInternal(std::shared_ptr<Layer> layer)
 {
-  s_layers[layer->GetLayer()] = std::move(layer);
+  const Config::LayerType layerType = layer->GetLayer();
+  {
+    WriteLock lock(s_layersRWLock);
+    s_layers[layerType] = std::move(layer);
```
Comment (lioncash) [inline Source/Core/Common/Config/Config.cpp]:
```suggestion
    s_layers.insert_or_assign(layer_type, std::move(layer));
```

File: Source/Core/Common/Config/Config.cpp [hunk 11]
```diff
@@ -5,51 +5,78 @@
...
 static u32 s_callback_guards = 0;
 
-Layers* GetLayers()
-{
-  return &s_layers;
-}
+// Mac supports shared_mutex since 10.12 and we're targeting 10.10,
+// so only use unique locks there...
+#if __APPLE__
+static std::mutex s_layersRWLock;
```
Comment (lioncash) [inline Source/Core/Common/Config/Config.cpp]: `s_layers_rw_lock`

We use `snake_case` for our naming conventions in new code.

File: Source/Core/Common/Config/Config.cpp (hunk 3 above)
Comment (CookiePLMonster) [reply Source/Core/Common/Config/Config.cpp]: I deliberately placed it here to construct it prior to locking the mutex.

File: Source/Core/Common/Config/Layer.h (hunk 8 above)
Comment (CookiePLMonster) [reply Source/Core/Common/Config/Layer.h]: TIL! I wasn't aware of this, great :)

File: Source/Core/Common/Config/Config.cpp (hunk 3 above)
Comment (lioncash) [reply Source/Core/Common/Config/Config.cpp]: For what purpose? It's a non-conditional assignment to nullptr.

**Comment by CookiePLMonster** [comment]
Done, also fixed `ClearCurrentRunLayer`, it was still using `std::make_unique` despite migrating to `shared_ptr`.

**Comment by leoetlino** [comment]
Won't your second change (not using std::optional) re-introduce an issue where calling DeleteKey will not actually delete settings from the underlying IniFile? (#6154)

**Comment by CookiePLMonster** [comment]
Hmmmm, good point - I focused only on what `Layer.cpp` was doing with those values but didn't check the rest. It's indeed possible that it'll reintroduce that, how do I verify that best?

**Comment by CookiePLMonster** [comment]
OK so I've been able to verify that both `Layer::DeleteKey` and `Layer::DeleteAllKeys` are **unused**, Dolphin compiles fine with them removed. Therefore, I see two ways out of this:

1) Remove deleting functionality as it's not needed now.
2) Partially revert my change to still use `std::optional<std::string>`, but **without** reverting exceessive use of `operator[]` which created entries even in getters.

**Comment by leoetlino** [comment]
I think 2 would be the best solution here; partially revert but keep the getter fixes. The Delete functions will likely be used as more parts of the codebase are migrated over to the new config system.

**Comment by CookiePLMonster** [comment]
I just noticed the commit message did not reflect the current state of this code - so I resubmitted with an updated message so git history isn't polluted by false claims.

**Pull request merged**